pymc3/gp/gp.py

```diff
@@ -843,98 +844,104 @@ def __init__(self, mean_func=Zero(), cov_funcs=[Constant(0.0)]):
             cov_total = given['gp'].cov_func
+            cov_totals = given['gp'].cov_funcs
             mean_total = given['gp'].mean_func
         else:
             cov_total = self.cov_func
+            cov_totals = self.cov_funcs
             mean_total = self.mean_func
-        if all(val in given for val in ['X', 'y', 'noise']):
-            X, y, noise = given['X'], given['y'], given['noise']
-            # if not isinstance(noise, Covariance):
-            #     noise = pm.gp.cov.WhiteNoise(noise)
+        if all(val in given for val in ['Xs', 'y', 'sigma']):
+            Xs, y, sigma = given['Xs'], given['y'], given['sigma']
         else:
-            X, y, noise = self.X, self.y, self.noise
-        return X, y, noise, cov_total, mean_total
+            Xs, y, sigma = self.Xs, self.y, self.sigma
+        self._check_inputs(Xs, y, sigma)
+        return Xs, y, sigma, cov_total, cov_totals, mean_total
 
-    def _build_conditional(self, Xnew, pred_noise, diag, Xs, y, noise,
-                           cov_total, mean_total):
+    def _build_conditional(self, Xnew, pred_noise, diag, Xs, y, sigma,
+                           cov_total, cov_totals, mean_total):
         # Old points
-        delta = y - self.mean_func(cartesian(*Xs))
-        Kns = [f(X) for f, X in zip_longest(self.cov_funcs, Xs)]
-        eigs_sep, Qs = zip(*map(np.linalg.eigh, Kns))  # Unzip
-        QTs = list(map(np.transpose, Qs))
+        X = cartesian(*Xs)
+        delta = y - mean_total(X)
+        Kns = [f(x) for f, x in zip(cov_totals, Xs)]
+        eigs_sep, Qs = zip(*map(eigh, Kns))  # Unzip
+        QTs = list(map(tt.transpose, Qs))
         eigs = kron_diag(eigs_sep)  # Combine separate eigs
-        if noise is not None:
-            eigs += noise
+        if sigma is not None:
+            eigs += sigma
 
         # New points
-        Km = self.total_cov_func(Xnew, diag)
-        Knm = self.total_cov_func(cartesian(*Xs), Xnew)
+        Km = cov_total(Xnew, diag)
+        Knm = cov_total(X, Xnew)
         Kmn = Knm.T
 
         # Build conditional mu
         alpha = kron_dot(QTs, delta)
-        alpha = alpha/self.eigs[:, None]
+        alpha = alpha/eigs[:, None]
         alpha = kron_dot(Qs, alpha)
-        mu = tt.dot(Kmns, alpha) + self.mean_func(Xnew)
+        mu = tt.dot(Kmns, alpha) + mean_total(Xnew)
 
         # Build conditional cov
         A = kron_dot(QTs, Knm)
-        A = A/tt.sqrt(self.eigs[:, None])
+        A = A/tt.sqrt(eigs[:, None])
         if diag:
             Asq = tt.sum(tt.square(A), 0)
             cov = Km - Asq
             if pred_noise:
-                cov += noise
+                cov += sigma
         else:
             Asq = tt.dot(A.T, A)
             cov = Km - Asq
             if pred_noise:
-                cov += noise*np.eye(cov.shape)
+                cov += sigma * np.eye(cov.shape)
         return mu, cov
 
     def conditional(self, name, Xnew, pred_noise=False, given=None, **kwargs):
```

Comment:
What do you think of adding a `predict` method?  Similar to the `Marginal` class.  The predict method is there as a convenience when using the MAP point to do predictions.

Reply:
Yes, it's on its way!

Reply:
nice!